### DISCUSSION TITLE
Highlight incorrect commands in red

### DISCUSSION THREAD

**Discussion by txtyash** [nushell/nushell#5238] [Ideas]
In fish the incorrect commands are highlighted in red and the correct ones in green or blue.
example:
If `porc` command doesn't exists then fish will highlight it with a different color(usually red) but if it exists then it'll be in blue or green.

**Comment by kubouch** [comment]
This should be possible but the result will be, at least with how it works currently, that every command will be highlighted in red during typing until it matches some known command which would be annoying. Also, checking for external command every keystroke might be quite expensive.

That being said, I guess it can be done, maybe ti could be a config setting.

**Reply by txtyash** [reply]
I think you shouldn't be checking for every key press. Use black(or white) color while the command is being entered and when the user hits space, check if exists(blue) or not(red).

**Reply by xosxos** [reply]
> every command will be highlighted in red during typing until it matches some known command which would be annoying. Also, checking for external command every keystroke might be quite expensive.

My ZSH syntax highlighting works exactly like this and I feel it is so satisfying seeing the commands turn green that it's one of the only things holding me back from switching at the moment :)

**Comment by DejayRezme** [comment]
I do like the idea of commands having a different syntax coloring depending on whether they are internal or external or not found. And different colors for language keywords like `let` `if` and `for` too.

I've noticed also that the syntax checking during typing can be a bit distracting with some statements. Maybe there should be some kind of delay, like syntax is only checked 200 or 500 milliseconds after the last keystroke. Not sure how IDEs handle incomplete statements with auto completion. 

For example when you type "let a =" the = is painted red until you continue typing, but it makes it feel like you made an error that you should correct before continuing.
Meanwhile parameter of internal commands are not syntax checked before pressing enter.

**Reply by zoechi** [reply]
Currently nu is already highlighting parts of the command in red while typing for example 
![Screenshot_20230202_163618](https://user-images.githubusercontent.com/405837/216370347-ac7df54f-e486-4eeb-bbd3-8148ee6bb560.png)

**Comment by Lindenk** [comment]
This is also how my zsh prompt works. If a command doesn't exist, it's red while typing and turns green when it's a valid command. Is there any way to do this currently? Perhaps with some kind of hook on `shape_external` coloring?

**Reply by fdncred** [reply]
seems like we'd need a `shape_unknown` and call `which` on every keystroke in the command position to see if it exists or not. Seems like it may be a performance issue?

**Reply by Lindenk** [reply]
So far my zsh scripts handle it just fine. I think performance wouldn't be an issue if the hook is implemented with a trie or some other efficient data structure for lookups. Even a handful of ms for a lookup on keystroke wouldn't be noticeable

**Reply by HTofi** [reply]
I don't understand why this is supposed to be such a performance issue. As it is right now, Nushell already recognizes builtin commands and displays them in bold font when you type, while non-builtin commands are shown in regular font. The font switches between regular and bold on a single key stroke, so extending this feature to color invalid commands in red shouldn't be that problematic, right?

**Reply by fdncred** [reply]
The problem is not distinguishing built-in vs not-built-in. If that were the case, you could just change this setting in your theme `shape_external: red`. The problem is determining whether the non-built-in is really an external in your path or just something not recognized.

However, if you don't care about "real" external executables, then that setting in your them may be a good compromise.

**Comment by Lindenk** [comment]
I ran a quick test with hyperfine on my system. The absolute least efficient but easiest method of checking if an executable is in your path is just running `which` each keystroke. According to hyperfine it takes an average of **0.2ms** to run. I think it's safe to say this isn't a performance issue

**Reply by fdncred** [reply]
That all depends on what OS you're on. Linux is fast. Windows is slow. MacOS is in the middle. WSL with windows paths is intolerably slow.

**Reply by Lindenk** [reply]
Sounds like it would make a great opt in config option then

**Comment by viperML** [comment]
This is one of the features that holds me back from moving to nushell from fish/zsh

**Reply by fdncred** [reply]
It was implemented in the last release with this setting https://github.com/nushell/nushell/blob/cd0a52cf00a508476574d94f66b0841387df1631/crates/nu-utils/src/sample_config/default_config.nu#L239 and this is where you choose the color https://github.com/nushell/nushell/blob/cd0a52cf00a508476574d94f66b0841387df1631/crates/nu-utils/src/sample_config/default_config.nu#L110 for the light_theme.

**Comment by thallium** [comment]
It looks like aliases to external commands don't get resolved. Is that intended?

**Reply by fdncred** [reply]
They do for me.

Not resolved because it doesn't exist in the path.
![image](https://github.com/nushell/nushell/assets/343840/5158afd0-932e-4a7b-b5ac-c718a5fa8962)
Resolved because it does exist in the path.
![image](https://github.com/nushell/nushell/assets/343840/07fc17ae-34d1-488c-a0b8-1eeb25b4d5fb)
Internal commands resolved.
![image](https://github.com/nushell/nushell/assets/343840/b3944ff4-9085-4410-bb9b-e5d2ab30571e)

**Reply by thallium** [reply]
I have `nv` aliased to `nvim` but `nv` doesn't get resolved while `nvim` is resolved:
<img width="484" alt="image" src="https://github.com/nushell/nushell/assets/20818480/bfe51836-be20-49e7-a3ee-824e618afee7">
<img width="382" alt="image" src="https://github.com/nushell/nushell/assets/20818480/3bfedf54-ea20-4d76-9c9e-d1ece64b8484">

**Reply by fdncred** [reply]
I'm not sure why aliases don't resolve. It could be some implementation detail with the `which` crate. All that nushell is doing is it's calling which on every character you type.

**Reply by thallium** [reply]
I took a quick look at the code. Looks like `highlight_resolved_externals` only applies to externals and aliases are not externals (didn't realize this before). And looks like aliases to externals are never resolved?
